File: topcoder/1259/zigzag.java
public class ZigZag {
    public int longestZigZag(int[] sequence) {
        int n = sequence.length;

        int[][] z = new int[n][2];
        z[0][0] = z[0][1] = 1;

        int best = 1;

        for (int i = 1; i < n; i++) {
            for (int j = i - 1; j >= 0; j--) {
                if (sequence[j] < sequence[i])
                    z[i][0] = Math.max(z[j][1] + 1, z[i][0]);
                else if (sequence[j] > sequence[i])
                    z[i][1] = Math.max(z[j][0] + 1, z[i][1]);
            }
            best = Math.max(best, Math.max(z[i][0], z[i][1]));
        }
        return best;
    }
}

File: topcoder/1259/zigzag.cpp
#include <string>
#include <vector>
#include <sstream>
#include <algorithm>
#include <iostream>
 
using namespace std;
 
class ZigZag {
  public:
  int longestZigZag(vector <int> sequence)  {
    if(sequence.size() == 1) return 1;
    vector<int> v(sequence.size()-1);
    for(int i=1;i<sequence.size();i++) 
      v[i-1] = sequence[i] - sequence[i-1];
    int ii=0;
    while( ii < v.size() && v[ii] == 0)
    ii++;
    if(ii == v.size()) return 1;
    int dir = v[ii]; 			// dir is first nonzero
    int len = 2;
    for(int i=ii+1;i<v.size();i++)
      if(v[i] * dir < 0) {
        dir *= -1;  
        len++;
      }
    return len;
  }
};